Sources/SwiftDocC/DocumentationService/Convert/ConvertService.swift

```diff
@@ -177,62 +137,102 @@ public struct ConvertService: DocumentationService {
+                dataProvider = FileManager.default
+            } else {
+                (bundle, dataProvider) = Self.makeBundleAndInMemoryDataProvider(request)
+            }
+            
+            let context = try DocumentationContext(bundle: bundle, dataProvider: dataProvider, configuration: configuration)
+            
+            // Precompute the render context
+            let renderContext = RenderContext(documentationContext: context, bundle: bundle)
+            
+            let symbolIdentifiersMeetingRequirementsForExpandedDocumentation: [String]? = request.symbolIdentifiersWithExpandedDocumentation?.compactMap { identifier, expandedDocsRequirement in
+                guard let documentationNode = context.documentationCache[identifier] else {
+                    return nil
+                }
+                
+                return documentationNode.meetsExpandedDocumentationRequirements(expandedDocsRequirement) ? identifier : nil
+            }
+            let converter = DocumentationContextConverter(
+                bundle: bundle,
+                context: context,
+                renderContext: renderContext,
                 emitSymbolSourceFileURIs: request.emitSymbolSourceFileURIs,
                 emitSymbolAccessLevels: true,
-                symbolIdentifiersWithExpandedDocumentation: request.symbolIdentifiersWithExpandedDocumentation
+                sourceRepository: nil,
+                symbolIdentifiersWithExpandedDocumentation: symbolIdentifiersMeetingRequirementsForExpandedDocumentation
             )
-
-            // Run the conversion.
-
-            let outputConsumer = OutputConsumer()
-            let (_, conversionProblems) = try converter.convert(outputConsumer: outputConsumer)
-
-            guard conversionProblems.isEmpty else {
-                throw ConvertServiceError.conversionError(
-                    underlyingError: DiagnosticConsoleWriter.formattedDescription(for: conversionProblems))
+            
+            let referencesToConvert: [ResolvedTopicReference]
+            if request.documentPathsToConvert == nil && request.externalIDsToConvert == nil {
+                // Should build all symbols
+                referencesToConvert = context.knownPages
+            }
+            else {
+                let symbolReferencesToConvert = Set(
+                    (request.externalIDsToConvert ?? []).compactMap { context.documentationCache.reference(symbolID: $0) }
+                )
+                let documentPathsToConvert = request.documentPathsToConvert ?? []
+                
+                referencesToConvert = context.knownPages.filter {
+                    symbolReferencesToConvert.contains($0) || documentPathsToConvert.contains($0.path)
+                }
+            }
+            
+            // Accumulate the render nodes
+            let renderNodes: [RenderNode] = referencesToConvert.concurrentPerform { reference, results in
```

Comment:
See the comment about this loop over known pages being written twice.

Reply:
See my [reply above](https://github.com/swiftlang/swift-docc/pull/1057#discussion_r1814544385)